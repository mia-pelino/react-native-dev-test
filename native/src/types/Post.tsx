export interface Author {
  id: string;
  name: string;
}

export interface Post {
  author: Author;
  body: string;
  id: string;
  publishedAt: Date;
  title: string;
}
