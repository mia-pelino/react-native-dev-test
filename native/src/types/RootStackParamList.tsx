export type RootStackParamList = {
  Home: undefined;
  FullPost: {
    title: string;
    body: string;
  };
};
