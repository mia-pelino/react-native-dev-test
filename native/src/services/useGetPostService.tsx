import { useEffect, useState } from 'react';
import { Post } from '../types/Post';
import { Service } from '../types/Service';

const useGetPostService = (setPostHandler: (posts: Post[]) => void) => {
  const [result, setResult] = useState<Service<Post[]>>({
    status: 'loading',
  });

  useEffect(() => {
    fetch('http://10.0.0.21:4000/posts')
      .then((response) => response.json())
      .then((response) =>
        setResult({
          status: 'loaded',
          payload: sortPosts(response, setPostHandler),
        })
      )
      .catch((error) => setResult({ status: 'error', error }));
  }, []);

  return result;
};

export default useGetPostService;

export const sortPosts = (
  posts: Post[],
  setPostHandler: (posts: Post[]) => void
) => {
  const sortedPosts = sortPostsReverseChronologically(posts);
  setPostHandler(sortedPosts);
  return sortedPosts;
};

export const sortPostsReverseChronologically = (posts: Post[]): Post[] => {
  return [].slice
    .call(posts)
    .sort(
      (newer: Post, older: Post): number =>
        new Date(older.publishedAt).getTime() -
        new Date(newer.publishedAt).getTime()
    );
};
