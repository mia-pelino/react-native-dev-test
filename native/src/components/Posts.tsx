import React, { useState } from 'react';
import { StackNavigationProp } from '@react-navigation/stack';
import { StyleSheet, Text, View, ActivityIndicator } from 'react-native';
import useGetLatinPostService from '../services/useGetPostService';
import PostList from './PostList';
import { Post } from '../types/Post';
import { RootStackParamList } from '../types/RootStackParamList';

type PostsScreenNavigationProp = StackNavigationProp<
  RootStackParamList,
  'Home'
>;

type Props = {
  navigation?: PostsScreenNavigationProp;
};

const Posts: React.FC<Props> = ({ navigation }) => {
  const [posts, setPosts] = useState<Post[]>([]);
  const [unfilteredPosts, setUnfilteredPosts] = useState<Post[]>(posts);
  const [isFiltered, setIsFiltered] = useState<boolean>(false);

  const service = useGetLatinPostService(setPosts);

  const filterByAuthor = (name: string): void => {
    setUnfilteredPosts(posts);
    setPosts(posts.filter((post) => post.author.name == name));
    setIsFiltered(true);
  };

  const displayAllPosts = (): void => {
    setIsFiltered(false);
    setPosts(unfilteredPosts);
  };

  const navigateToFullPost = (title: string, body: string): void => {
    navigation?.navigate('FullPost', { title: title, body: body });
  };

  return (
    <View style={styles.container}>
      {service.status === 'loading' && <ActivityIndicator />}
      {service.status === 'loaded' && service.payload.length > 0 && (
        <PostList
          postList={posts}
          isFiltered={isFiltered}
          pressAuthorHandler={(name) => {
            filterByAuthor(name);
          }}
          pressPostHandler={(title, body) => navigateToFullPost(title, body)}
          displayAllPostsHandler={() => displayAllPosts()}
        />
      )}
      {service.status === 'error' && <Text>Oops, I did it again.</Text>}
    </View>
  );
};

export default Posts;

const styles = StyleSheet.create({
  container: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center',
    backgroundColor: '#000',
  },
});
