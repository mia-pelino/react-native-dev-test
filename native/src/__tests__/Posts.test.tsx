import React from 'react';
import renderer, { create, act } from 'react-test-renderer';
import { mocked } from 'ts-jest/utils';
import 'isomorphic-fetch';
import useGetPostService from '../services/useGetPostService';
import { Service } from '../types/Service';
import { Post } from '../types/Post';
import Posts from '../components/Posts';

jest.mock('../services/useGetPostService');
const mockedService = mocked(useGetPostService, true);

describe('<Posts /> ', () => {
  describe('snapshots', () => {
    it('shows loading screen while service status is still loading', async () => {
      const expectedResult: Service<Post[]> = {
        status: 'loading',
      };
      mockedService.mockReturnValueOnce(expectedResult);

      const tree = renderer.create(<Posts />).toJSON();

      await act(async () => {
        expect(tree).toMatchSnapshot();
      });
    });

    it('shows error screen for service status error', async () => {
      const expectedResult: Service<Post[]> = {
        status: 'error',
        error: new Error(),
      };
      mockedService.mockReturnValueOnce(expectedResult);

      const tree = renderer.create(<Posts />).toJSON();

      await act(async () => {
        expect(tree).toMatchSnapshot();
      });
    });

    it('shows post list screen when results have loaded', async () => {
      const expectedResult: Service<Post[]> = {
        status: 'loaded',
        payload: [],
      };
      mockedService.mockReturnValueOnce(expectedResult);

      const tree = renderer.create(<Posts />).toJSON();
      await act(async () => {
        expect(tree).toMatchSnapshot();
      });
    });
  });

  describe('component', () => {
    it('renders ActivityIndicator component during status of loading', () => {
      const expectedResult: Service<Post[]> = {
        status: 'loading',
      };
      mockedService.mockReturnValueOnce(expectedResult);

      const tree = renderer.create(<Posts />).toJSON().children[0];

      expect(tree.type).toBe('ActivityIndicator');
    });

    it('displays error message when loaded with error', () => {
      const expectedResult: Service<Post[]> = {
        status: 'error',
        error: new Error(),
      };
      mockedService.mockReturnValueOnce(expectedResult);

      const tree = renderer.create(<Posts />).toJSON().children[0];

      expect(tree.type).not.toBe('ActivityIndicator');
      expect(tree.children[0]).toBe('Oops, I did it again.');
    });

    it('displays post list when results have loaded', () => {
      const postData = [
        {
          title: 'tater tots',
          body: 'po \n\n## ta **toes**',
          author: {
            name: 'samwise',
            id: '1',
          },
          id: 'a',
          publishedAt: new Date('01-02-2020'),
        },
      ];

      const expectedResult: Service<Post[]> = {
        status: 'loaded',
        payload: postData,
      };
      mockedService.mockReturnValueOnce(expectedResult);

      const tree = renderer.create(<Posts />).toJSON().children[0].children[0];

      expect(tree.children[0]).toBe(
        `Press an author's name to only show their posts.`
      );
    });
  });
});
