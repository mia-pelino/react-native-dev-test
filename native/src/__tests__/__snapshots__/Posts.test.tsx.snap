// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Posts />  snapshots shows error screen for service status error 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#000",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <Text>
    Oops, I did it again.
  </Text>
</View>
`;

exports[`<Posts />  snapshots shows loading screen while service status is still loading 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#000",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <ActivityIndicator
    animating={true}
    color="#999999"
    hidesWhenStopped={true}
    size="small"
  />
</View>
`;

exports[`<Posts />  snapshots shows post list screen when results have loaded 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#000",
      "flex": 1,
      "justifyContent": "center",
    }
  }
/>
`;
